### ISSUE TITLE
Redundant connections in share mode

### ISSUE THREAD

**Issue by RicoTilgner** [arvidn/libtorrent#3223]
libtorrent version (or branch): master

platform/architecture: x64, linux

compiler and compiler version: gcc 8.1

When having a lot of torrents in share mode my client is often staying connected to many seeds for torrents that don't have any leechers, taking up connections for torrents where there actually are leechers.
I dug a bit through the code and found in https://github.com/arvidn/libtorrent/blob/master/src/peer_connection.cpp that share_mode disables close_redundant_connections. I get that this is good when leechers are present, since you might want to get the next piece. But in cases where there are no leechers I think it should not try to hold a ton of connections open. Wdyt?

**Comment by arvidn** [comment]
you mean that in share-mode there should be a distinction between redundant-connections (as traditionally defined) and connections with other share-mode peers?

i.e. the case you're concerned about is where there are a lot of other share-mode peers.

**Comment by RicoTilgner** [comment]
no, I'm not concerned about other share-mode clients, I'm concerned that the very finite number of ports that I have is used by seeders on torrents that don't have any leechers and I can't connect to the leechers on the few torrents that have some.
Right now a share-mode torrent will permanently keep 20ish connections to seeders open, mainly to determine which piece to load next. When having 2000 torrents in share-mode this would be 40000 connections. :)
I changed the share-mode code locally a little to not require seeders to be connected anymore and enabled closing of redundant connections. This is working quite well except for torrents that are currently super-seeding or where a full copy isn't available. (I cut a few corners when deciding which piece to load next)
I also disabled the requirement that a piece that I request has to be requested by a certain number of leechers at the same time but rather only rely on the ratio. This helps with torrents that don't have many leechers and where I'll only see one leecher at a time.
